### PR TITLE
Turn warnings as errors  for doxygen and add missing docs.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/Azure/azure-sdk-for-c/issues/1064

### CODE REVIEW

File: Doxyfile [hunk 1]
```diff
@@ -811,7 +811,7 @@ WARN_FORMAT            = "$file:$line: $text"
 # messages should be written. If left blank the output is written to standard
```
Comment (weshaggard) [inline Doxyfile:811]: orthogonal to your change but I'm curious if this file can be moved under the eng directory to help cut down on the pollution in the root repo.

Comment (ahsonkhan) [reply Doxyfile:811]: The actual warning file (i.e. `DoxygenWarningLog.txt`), or this `Doxyfile` file, or both?

I can certainly move the warning file to  `eng\DoxygenWarningLog.txt`

Comment (weshaggard) [reply Doxyfile:811]: I was referencing the DoxyFile but the warning file should probably go wherever we put normal logs for other build logs.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 2]
```diff
@@ -338,7 +338,15 @@ jobs:
...
-    - script: '"C:/Program Files/doxygen/bin/doxygen.exe" Doxyfile'
+      # If doxygen produces any warnings, fail the step.
+    - pwsh: |
+        & 'C:\Program Files\doxygen\bin\doxygen.exe' Doxyfile
+        if (Test-Path "DoxygenWarningLog.txt" -PathType leaf)
+        {
+          Get-Content "DoxygenWarningLog.txt"
+          Write-Error "Warnings from the doxygen docs generation step are treated as error."
+          exit 1
+        }
```
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:349]: @weshaggard, @danieljurek - please review this change.

File: Doxyfile (hunk 1 above)
Comment (ahsonkhan) [reply Doxyfile:811]: > orthogonal to your change but I'm curious if this file can be moved under the eng directory to help cut down on the pollution in the root repo.

https://github.com/Azure/azure-sdk-for-c/issues/1146

**Pull request merged**